File: sdk-java/src/test/java/ng/linkbridge/sdk/ClientTest.java
package ng.linkbridge.sdk;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertNotEquals;
import static org.junit.jupiter.api.Assertions.assertNull;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.time.Instant;
import java.time.InstantSource;
import java.util.List;
import java.util.Map;
import java.util.concurrent.atomic.AtomicLong;
import org.junit.jupiter.api.Test;

class ClientTest {

  private static LinkbridgeClient.Builder base(FakeTransport t) {
    return LinkbridgeClient.builder()
        .baseUrl("https://api.example.test")
        .clientId("lb_test_abc")
        .clientSecret("secret")
        .transport(t);
  }

  // ── config validation ──────────────────────────────────────────────────

  @Test
  void baseUrlIsRequired() {
    IllegalArgumentException e =
        assertThrows(
            IllegalArgumentException.class,
            () -> LinkbridgeClient.builder().clientId("a").clientSecret("b").build());
    assertTrue(e.getMessage().contains("base_url is required"));
  }

  @Test
  void credentialsOrStaticTokenRequired() {
    IllegalArgumentException e =
        assertThrows(
            IllegalArgumentException.class,
            () -> LinkbridgeClient.builder().baseUrl("https://api.example.test").build());
    assertTrue(e.getMessage().contains("static_token or client_id+client_secret"));
  }

  @Test
  void trailingSlashesStripped() {
    FakeTransport t = new FakeTransport();
    LinkbridgeClient c =
        LinkbridgeClient.builder()
            .baseUrl("https://api.example.test///")
            .staticToken("tok")
            .transport(t)
            .build();
    c.request("GET", "/v1/invoices");
    assertEquals("https://api.example.test/v1/invoices", t.last().uri().toString());
  }

  // ── user agent ─────────────────────────────────────────────────────────

  @Test
  void userAgentDefaultsAndSuffix() {
    FakeTransport t = new FakeTransport();
    LinkbridgeClient plain = base(t).build();
    assertEquals("linkbridge-java/" + LinkbridgeClient.SDK_VERSION, plain.userAgentForTesting());

    LinkbridgeClient suffixed = base(t).userAgent("myapp/1.0").build();
    assertTrue(suffixed.userAgentForTesting().startsWith("linkbridge-java/"));
    assertTrue(suffixed.userAgentForTesting().endsWith(" myapp/1.0"));
  }

  // ── token flow ─────────────────────────────────────────────────────────

  @Test
  void tokenFetchedOnceAndCached() {
    FakeTransport t = new FakeTransport();
    LinkbridgeClient c = base(t).build();
    c.request("GET", "/v1/invoices");
    c.request("GET", "/v1/invoices");
    assertEquals(1, t.tokenCalls);
    assertEquals("Bearer tok-1", t.lastApi().headers().get("authorization"));
  }

  @Test
  void tokenRequestShape() {
    FakeTransport t = new FakeTransport();
    base(t).scopes(List.of("invoices:write", "webhooks:manage")).build().request("GET", "/v1/x");
    FakeTransport.Recorded tok = t.requests.get(0);
    assertEquals("/v1/oauth/token", tok.uri().getPath());
    assertEquals("POST", tok.method());
    assertNull(tok.headers().get("authorization"), "token request must carry no bearer");
    String body = tok.bodyText();
    assertTrue(body.contains("\"grant_type\":\"client_credentials\""));
    assertTrue(body.contains("\"client_id\":\"lb_test_abc\""));
    assertTrue(body.contains("\"scope\":\"invoices:write webhooks:manage\""), body);
  }

  @Test
  void defaultScopes() {
    FakeTransport t = new FakeTransport();
    base(t).build().request("GET", "/v1/x");
    assertTrue(t.requests.get(0).bodyText().contains("\"scope\":\"invoices:write invoices:read\""));
  }

  @Test
  void tokenRefreshesInsideSixtySecondWindow() {
    FakeTransport t = new FakeTransport();
    t.tokenResponse = "{\"access_token\":\"tok-1\",\"expires_in\":120}";
    LinkbridgeClient c = base(t).build();
    AtomicLong now = new AtomicLong(1_700_000_000L);
    c.setClockForTesting(InstantSource.fixed(Instant.ofEpochSecond(now.get())));

    c.request("GET", "/v1/x");
    assertEquals(1, t.tokenCalls);

    // 59s of validity left (<= 60) → must refresh.
    c.setClockForTesting(InstantSource.fixed(Instant.ofEpochSecond(now.get() + 61)));
    c.request("GET", "/v1/x");
    assertEquals(2, t.tokenCalls);
  }

  @Test
  void missingExpiresInFallsBackTo300() {
    FakeTransport t = new FakeTransport();
    t.tokenResponse = "{\"access_token\":\"tok-1\"}";
    LinkbridgeClient c = base(t).build();
    c.setClockForTesting(InstantSource.fixed(Instant.ofEpochSecond(1_700_000_000L)));
    c.request("GET", "/v1/x");
    // 300s fallback: at +200s there are 100s left (> 60) → still cached.
    c.setClockForTesting(InstantSource.fixed(Instant.ofEpochSecond(1_700_000_200L)));
    c.request("GET", "/v1/x");
    assertEquals(1, t.tokenCalls);
    // at +250s only 50s left → refresh.
    c.setClockForTesting(InstantSource.fixed(Instant.ofEpochSecond(1_700_000_250L)));
    c.request("GET", "/v1/x");
    assertEquals(2, t.tokenCalls);
  }

  @Test
  void staticTokenNeverCallsTokenEndpoint() {
    FakeTransport t = new FakeTransport();
    LinkbridgeClient c =
        LinkbridgeClient.builder()
            .baseUrl("https://api.example.test")
            .staticToken("static-tok")
            .transport(t)
            .build();
    c.request("GET", "/v1/invoices");
    c.request("GET", "/v1/webhooks");
    assertEquals(0, t.tokenCalls);
    assertEquals("Bearer static-tok", t.lastApi().headers().get("authorization"));
  }

  @Test
  void tokenRefreshIsSingleFlightUnderConcurrency() throws Exception {
    FakeTransport t = new FakeTransport();
    LinkbridgeClient c = base(t).build();
    int threads = 12;
    var pool = java.util.concurrent.Executors.newFixedThreadPool(threads);
    var start = new java.util.concurrent.CountDownLatch(1);
    var done = new java.util.concurrent.CountDownLatch(threads);
    for (int i = 0; i < threads; i++) {
      pool.submit(
          () -> {
            try {
              start.await();
              c.request("GET", "/v1/invoices");
            } catch (Exception ignored) {
              // fall through — the assertion below catches failures
            } finally {
              done.countDown();
            }
          });
    }
    start.countDown();
    assertTrue(done.await(10, java.util.concurrent.TimeUnit.SECONDS));
    pool.shutdown();
    assertEquals(1, t.tokenCalls, "concurrent callers must coalesce onto one token fetch");
  }

  @Test
  void emptyAccessTokenIsInvalidTokenResponse() {
    FakeTransport t = new FakeTransport();
    t.tokenResponse = "{\"access_token\":\"\",\"expires_in\":3600}";
    LinkbridgeApiException e =
        assertThrows(LinkbridgeApiException.class, () -> base(t).build().request("GET", "/v1/x"));
    assertEquals("invalid_token_response", e.code());
    assertEquals("empty access_token", e.apiMessage());
  }

  // ── idempotency key ────────────────────────────────────────────────────

  @Test
  void idempotencyKeyFormatAndUniqueness() {
    String a = LinkbridgeClient.idempotencyKey();
    String b = LinkbridgeClient.idempotencyKey();
    assertTrue(a.matches("^lb-[0-9a-f]{32}$"), a);
    assertTrue(b.matches("^lb-[0-9a-f]{32}$"), b);
    assertNotEquals(a, b);
  }

  // ── query + path encoding ──────────────────────────────────────────────

  @Test
  void nullAndEmptyQueryValuesDropped() {
    FakeTransport t = new FakeTransport();
    t.enqueue(200, "{\"data\":[],\"next_cursor\":null}");
    base(t).build().invoices().list(null, 20, "");
    String q = t.lastApi().uri().getQuery();
    assertEquals("limit=20", q);
    assertFalse(t.lastApi().uri().toString().contains("cursor="));
    assertFalse(t.lastApi().uri().toString().contains("status="));
  }

  @Test
  void pathSegmentsFullyPercentEncoded() {
    assertEquals("INV%2F1%20x", LinkbridgeClient.encodePathSegment("INV/1 x"));
    FakeTransport t = new FakeTransport();
    t.enqueue(200, "{\"irn\":\"INV/1\",\"status\":\"pending\"}");
    base(t).build().invoices().get("INV/1");
    assertTrue(t.lastApi().uri().getRawPath().endsWith("/v1/invoices/INV%2F1"));
  }

  // ── response + error decoding ──────────────────────────────────────────

  @Test
  void errorEnvelopeDecoded() {
    FakeTransport t = new FakeTransport();
    t.enqueue(
        422,
        "{\"error\":{\"code\":\"schema_validation_failed\",\"message\":\"invoice payload does not satisfy the canonical schema\",\"details\":[\"/irn: bad\"],\"trace_id\":\"abc/def-000001\"}}");
    LinkbridgeApiException e =
        assertThrows(LinkbridgeApiException.class, () -> base(t).build().request("GET", "/v1/x"));
    assertEquals(422, e.status());
    assertEquals("schema_validation_failed", e.code());
    assertEquals("abc/def-000001", e.traceId());
    assertEquals("/irn: bad", e.details().get(0).asText());
    assertEquals(
        "linkbridge: 422 schema_validation_failed: invoice payload does not satisfy the canonical schema",
        e.getMessage());
  }

  @Test
  void nonJsonErrorBodyFallsBackToHttpError() {
    FakeTransport t = new FakeTransport();
    t.enqueue(502, "<html>Bad Gateway</html>");
    LinkbridgeApiException e =
        assertThrows(LinkbridgeApiException.class, () -> base(t).build().request("GET", "/v1/x"));
    assertEquals("http_error", e.code());
    assertEquals("<html>Bad Gateway</html>", e.apiMessage());
  }

  @Test
  void emptyErrorBodyMessageIsHttpStatus() {
    FakeTransport t = new FakeTransport();
    t.enqueue(503, "");
    LinkbridgeApiException e =
        assertThrows(LinkbridgeApiException.class, () -> base(t).build().request("GET", "/v1/x"));
    assertEquals("http_error", e.code());
    assertEquals("http 503", e.apiMessage());
  }

  @Test
  void nonJsonTwoxxBodyIsInvalidJsonResponse() {
    FakeTransport t = new FakeTransport();
    t.enqueue(200, "not-json");
    LinkbridgeApiException e =
        assertThrows(LinkbridgeApiException.class, () -> base(t).build().request("GET", "/v1/x"));
    assertEquals("invalid_json_response", e.code());
  }

  @Test
  void emptyBodyReturnsNull() {
    FakeTransport t = new FakeTransport();
    t.queue.add(new Transport.Response(204, Map.of(), new byte[0]));
    assertNull(base(t).build().request("DELETE", "/v1/webhooks/x"));
  }

  @Test
  void contentTypeOnlyWhenBodyPresent() {
    FakeTransport t = new FakeTransport();
    LinkbridgeClient c = base(t).build();
    c.request("GET", "/v1/invoices");
    assertNull(t.lastApi().headers().get("content-type"));
    c.request("POST", "/v1/webhooks", null, Map.of("url", "https://x.example"), null);
    assertEquals("application/json", t.lastApi().headers().get("content-type"));
  }
}

File: sdk-php/tests/ClientTest.php
<?php

declare(strict_types=1);

namespace Linkbridge\Tests;

use Linkbridge\ApiError;
use Linkbridge\LinkbridgeClient;
use PHPUnit\Framework\TestCase;

final class ClientTest extends TestCase
{
    private function newClient(FakeTransport $transport): LinkbridgeClient
    {
        return new LinkbridgeClient([
            'base_url'      => 'https://api.example.test',
            'client_id'     => 'cid',
            'client_secret' => 'csecret',
            'transport'     => $transport,
        ]);
    }

    public function testRejectsConstructionWithoutCredentials(): void
    {
        $this->expectException(\InvalidArgumentException::class);
        new LinkbridgeClient(['base_url' => 'https://api.example.test']);
    }

    public function testStaticTokenSkipsOauthRoundTrip(): void
    {
        $t = new FakeTransport();
        $t->enqueue(200, json_encode(['data' => [], 'next_cursor' => null]));

        $client = new LinkbridgeClient([
            'base_url'     => 'https://api.example.test',
            'static_token' => 'static-bearer',
            'transport'    => $t,
        ]);
        $client->invoices->list();

        $this->assertCount(1, $t->calls);
        $this->assertSame('https://api.example.test/v1/invoices', $t->calls[0]['url']);
        $this->assertSame('Bearer static-bearer', $t->calls[0]['headers']['Authorization']);
    }

    public function testTokenIsCachedAcrossCalls(): void
    {
        $t = new FakeTransport();
        // First call: oauth token, then GET /v1/invoices, then GET again.
        $t->enqueue(200, json_encode(['access_token' => 'tok-1', 'expires_in' => 3600, 'token_type' => 'Bearer']));
        $t->enqueue(200, json_encode(['data' => [], 'next_cursor' => null]));
        $t->enqueue(200, json_encode(['data' => [], 'next_cursor' => null]));

        $client = $this->newClient($t);
        $client->invoices->list();
        $client->invoices->list();

        $this->assertCount(3, $t->calls, 'token call should be cached for the second list');
        $this->assertSame('POST', $t->calls[0]['method']);
        $this->assertStringEndsWith('/v1/oauth/token', $t->calls[0]['url']);
        $this->assertSame('Bearer tok-1', $t->calls[1]['headers']['Authorization']);
        $this->assertSame('Bearer tok-1', $t->calls[2]['headers']['Authorization']);
    }

    public function testSubmitInjectsIdempotencyKeyAndJsonHeaders(): void
    {
        $t = new FakeTransport();
        $t->enqueue(200, json_encode(['access_token' => 'tok', 'expires_in' => 3600]));
        $t->enqueue(202, json_encode(['irn' => 'IRN-1', 'status' => 'accepted']));

        $client = $this->newClient($t);
        $resp = $client->invoices->submit(['supplier' => ['tin' => '12345678-0001']]);

        $this->assertSame('IRN-1', $resp['irn']);
        $submit = $t->calls[1];
        $this->assertSame('POST', $submit['method']);
        $this->assertSame('https://api.example.test/v1/invoices', $submit['url']);
        $this->assertSame('application/json', $submit['headers']['Content-Type']);
        $this->assertArrayHasKey('Idempotency-Key', $submit['headers']);
        $this->assertMatchesRegularExpression('/^lb-[0-9a-f]{32}$/', $submit['headers']['Idempotency-Key']);
        $this->assertSame(['supplier' => ['tin' => '12345678-0001']], json_decode($submit['body'], true));
    }

    public function testUpdateStatusSendsSpecFieldNames(): void
    {
        $t = new FakeTransport();
        $t->enqueue(200, json_encode(['access_token' => 'tok', 'expires_in' => 3600]));
        $t->enqueue(200, json_encode(['irn' => 'IRN-1', 'payment_status' => 'PAID']));
        $t->enqueue(200, json_encode(['irn' => 'IRN-1', 'payment_status' => 'UNPAID']));

        $client = $this->newClient($t);
        $client->invoices->updateStatus('IRN-1', 'PAID', 'RCPT-1');

        $update = $t->calls[1];
        $this->assertSame('POST', $update['method']);
        $this->assertSame('https://api.example.test/v1/invoices/IRN-1/status', $update['url']);
        $this->assertSame(
            ['payment_status' => 'PAID', 'reference' => 'RCPT-1'],
            json_decode($update['body'], true),
        );

        $client->invoices->updateStatus('IRN-1', 'UNPAID');
        $this->assertSame(
            ['payment_status' => 'UNPAID'],
            json_decode($t->calls[2]['body'], true),
            'reference key must be omitted entirely when not provided',
        );
    }

    public function testQueryStringIsAssembledFromFilters(): void
    {
        $t = new FakeTransport();
        $t->enqueue(200, json_encode(['access_token' => 'tok', 'expires_in' => 3600]));
        $t->enqueue(200, json_encode(['data' => []]));

        $client = $this->newClient($t);
        $client->invoices->list(['status' => 'transmitted', 'limit' => 25]);

        $this->assertSame(
            'https://api.example.test/v1/invoices?status=transmitted&limit=25',
            $t->calls[1]['url'],
        );
    }

    public function testApiErrorEnvelopeIsDecoded(): void
    {
        $t = new FakeTransport();
        $t->enqueue(200, json_encode(['access_token' => 'tok', 'expires_in' => 3600]));
        $t->enqueue(422, json_encode([
            'error' => [
                'code'     => 'validation_error',
                'message'  => 'totals.payable mismatch',
                'trace_id' => 'trace-xyz',
                'details'  => [['field' => 'totals.payable', 'rule' => 'crossCheckTotals']],
            ],
        ]));

        $client = $this->newClient($t);
        try {
            $client->invoices->submit(['x' => 1]);
            $this->fail('expected ApiError');
        } catch (ApiError $e) {
            $this->assertSame(422, $e->status);
            $this->assertSame('validation_error', $e->errorCode);
            $this->assertSame('validation_error', $e->code());
            $this->assertSame('trace-xyz', $e->traceId);
            $this->assertStringContainsString('totals.payable mismatch', $e->getMessage());
            $this->assertIsArray($e->details);
        }
    }

    public function testDeleteHandlesNoContent(): void
    {
        $t = new FakeTransport();
        $t->enqueue(200, json_encode(['access_token' => 'tok', 'expires_in' => 3600]));
        $t->enqueue(204, '');

        $client = $this->newClient($t);
        $client->webhooks->delete('wh_123');
        $this->assertSame('DELETE', $t->calls[1]['method']);
        $this->assertStringEndsWith('/v1/webhooks/wh_123', $t->calls[1]['url']);
    }

    public function testIdempotencyKeyHelperShape(): void
    {
        $key = LinkbridgeClient::idempotencyKey();
        $this->assertMatchesRegularExpression('/^lb-[0-9a-f]{32}$/', $key);
        $this->assertNotSame($key, LinkbridgeClient::idempotencyKey());
    }
}
